### PR TITLE
net: add winc1500 driver from Atmel

### PR DESCRIPTION
**Pull request by MaxAgneni**
This is part of a patch series to add support for winc1500 wifi module

origin: extracted from Atmel ASF version 3.33.0

was: https://gerrit.zephyrproject.org/r/#/c/12356/

### CODE REVIEW

**Review by tbursztyka — APPROVED** [review]

File: drivers/wifi/Kconfig [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Kconfig - drivers configuration options
+
+#
+# Copyright (c) 2015 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+
+menu "WiFi Drivers"
```
Comment (tbursztyka) [inline drivers/wifi/Kconfig]: change this into a menuconfig WIFI directly

File: drivers/wifi/Kconfig [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+config WIFI
+	bool
+	prompt "add support for WIFI Drivers"
```
Comment (tbursztyka) [inline drivers/wifi/Kconfig]: and you can avoid the prompt by doing: bool "Add support for WiFi drivers"

File: drivers/wifi/Kconfig [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	default n
+
```
Comment (tbursztyka) [inline drivers/wifi/Kconfig]: and then here you would do:

if WIFI

File: drivers/wifi/Kconfig [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	depends on WIFI
+	default 80
+	help
+	  WiFi device driver initialization priority.
+	  Do not mess with it unless you know what you are doing.
+	  Note that the priority needs to be lower than the net stack
+	  so that it can start before the networking sub-system.
+
+source "drivers/wifi/winc1500/Kconfig"
+endmenu
```
Comment (tbursztyka) [inline drivers/wifi/Kconfig]: and here, instead of endmenu:

endif # WIFI

File: drivers/wifi/winc1500/bsp/include/nm_bsp.h [hunk 5]
```diff
@@ -0,0 +1,284 @@
+/**
+ *
+ * \file
+ *
+ * \brief WINC BSP API Declarations.
+ *
+ * Copyright (c) 2016 Atmel Corporation. All rights reserved.
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bsp/include/nm_bsp.h]: Sounds like this file should go somewhere in ext/  or?

What goes into all directories but ext, is made for zephyr so there is no such thing as redefining stuff like NULL, uints ... etc... that's useless: Zephyr provides all necessary bits and pieces.

File: drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ *
+ */
+typedef struct {
+	struct device            **gpios;
+	struct gpio_callback       gpio_cb;
+
+	struct device             *spi_dev;
+	int                        spi_slave;
+} sWINC1500;
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h]: For Zephyr code: there is _no_ camelcase. Never.
(if it's code which is in ext/ then sure, but nowhere else)

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+#include <board.h>
+#include <gpio.h>
+
+#include "bsp/include/nm_bsp.h"
+#include "common/include/nm_common.h"
+#include "conf_winc.h"
+
+//#include <misc/printk.h>
+//#define PRINT printk
+#define PRINT(x...)
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: we have been using sys_log.h for quite a while now. Check other drivers, it's very useful. Debugging level is then Kconfig based, and you have access to macros like SYS_LOG_ERR, SYS_LOG_INF, SYS_LOG_DBG ...

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+ */
+sint8 nm_bsp_init(void)
+{
+	PRINT("nm_bsp_init:\n");
+	gpfIsr = NULL;
+
+	/* Initialize chip IOs. */
+	winc1500.gpios = winc1500_configure_gpios();
+
+	winc1500.spi_dev = device_get_binding(CONFIG_WINC1500_SPI_DRV_NAME);
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: and if that fails?

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+//#include <misc/printk.h>
+//#define PRINT printk
+#define PRINT(x...)
+
+/**
+ *
+ */
+sWINC1500 winc1500;
+
+static tpfNmBspIsr gpfIsr;
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: so as I said earlier: no camelcase.

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+/**
+ *	@fn		nm_bsp_reset
+ *	@brief	Reset NMC1500 SoC by setting CHIP_EN and RESET_N signals low,
+ *           CHIP_EN high then RESET_N high
+ */
+void nm_bsp_reset(void)
+{
+	gpio_pin_write(winc1500.gpios[WINC1500_GPIO_IDX_CHIP_EN],
+			       CONFIG_WINC1500_GPIO_CHIP_EN, 0);
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: indentation is as follows:
foo(param1, ...
      param2, paramn);

so it always align on previous opened '(' + 1 space.

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+/*
+ *	@fn		nm_bsp_sleep
+ *	@brief	Sleep in units of mSec
+ *	@param[IN]	u32TimeMsec
+ *				Time in milliseconds
+ */
+void nm_bsp_sleep(uint32 u32TimeMsec)
+{
+	k_busy_wait(u32TimeMsec*1000);
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: so no camelcase, but change this to:

K_MS(time_msec) * USEC_PER_SEC

USEC_PER_SEC is found in include/sys_clock.h

File: drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+#include <gpio.h>
+#include <spi.h>
+
+#include "bsp/include/nm_bsp.h"
+#include "common/include/nm_common.h"
+#include "bus_wrapper/include/nm_bus_wrapper.h"
+#include "conf_winc.h"
+
+#include <misc/printk.h>
+#define SYS_LOG_ERR(x...)
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c]: why redefining this??

File: drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+	winc1500.gpios = winc1500_configure_gpios();
+
+	/* setup SPI device */
+	winc1500.spi_dev = device_get_binding(CONFIG_WINC1500_SPI_DRV_NAME);
+	if (!winc1500.spi_dev) {
+		SYS_LOG_ERR("spi device binding\n");
+		return -1;
+	}
+
+	/* spi config is a bit field with the following parts:
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c]: no need to re-document how spi is configured as all is in spi.h already

File: drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+	return M2M_SUCCESS;
+}
+
+#endif
+
+/*
+*	@fn		nm_bus_init
+*	@brief	Initialize the bus wrapper
+*	@return	M2M_SUCCESS in case of success and M2M_ERR_BUS_FAIL in case of failure
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c]: that does not seem to be the case. You are returning -1 in error.

File: drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+		case NM_BUS_IOCTL_W: {
+			tstrNmI2cDefault *pstrParam = (tstrNmI2cDefault *)pvParameter;
+
+			s8Ret = nm_i2c_write(pstrParam->pu8Buf, pstrParam->u16Sz);
+		}
+		break;
+		case NM_BUS_IOCTL_W_SPECIAL: {
+			tstrNmI2cSpecial *pstrParam = (tstrNmI2cSpecial *)pvParameter;
+
+			s8Ret = nm_i2c_write_special(pstrParam->pu8Buf1, pstrParam->u16Sz1, pstrParam->pu8Buf2, pstrParam->u16Sz2);
```
Comment (tbursztyka) [inline drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c]: 80 chars limit.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Where is the NET_DEVICE_INIT or at least DEVICE_INIT?

**Review by jukkar — CHANGES_REQUESTED** [review]
Please have a proper commit messages. So subject, body and signed-off etc.

File: drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+#define NM_EDGE_INTERRUPT       (1)
+
+#define NM_DEBUG				CONF_WINC_DEBUG
+#define NM_BSP_PRINTF			CONF_WINC_PRINTF
+
+/**
+ *
+ */
+typedef struct {
```
Comment (jukkar) [inline drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h]: No typedefs for structs.

File: drivers/wifi/winc1500/bsp/include/nm_bsp.h (hunk 5 above)
Comment (jukkar) [inline drivers/wifi/winc1500/bsp/include/nm_bsp.h]: I agree with Tomasz, these changes look like they belong to etx.

**Review by jukkar — CHANGES_REQUESTED** [review]

**Review by tbursztyka — CHANGES_REQUESTED** [review]
And please have a proper commit message.

Signed-off-by etc... and we don't care of former gerrit url.

File: drivers/wifi/winc1500/bsp/include/nm_bsp.h (hunk 5 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bsp/include/nm_bsp.h]: I left the function prototypes in this file and moved the other definitions in ext

File: drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h (hunk 16 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h]: removed

File: drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h (hunk 6 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bsp/include/nm_bsp_internal.h]: removed

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c (hunk 7 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: Done

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c (hunk 9 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: Done

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c (hunk 8 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: Added spi_dev check

File: drivers/wifi/winc1500/bsp/source/nm_bsp.c (hunk 11 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bsp/source/nm_bsp.c]: Done

File: drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c (hunk 12 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c]: Removed

File: drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c (hunk 13 above)
Comment (MaxAgneni) [inline drivers/wifi/winc1500/bus_wrapper/source/nm_bus_wrapper.c]: removed and configured with spi defines

File: boards/x86/panther/Kconfig.defconfig [hunk 17]
```diff
@@ -4,6 +4,25 @@ if BOARD_PANTHER
 config BOARD
 	default "panther"
 
+if NBLE
```
Comment (nashif) [inline boards/x86/panther/Kconfig.defconfig]: Remove this, this board does not use NBLE, NBLE is only being used for Arduino 101 with a custom Bluetooth firmware

File: boards/x86/panther/Kconfig.defconfig [hunk 18]
```diff
@@ -4,6 +4,25 @@ if BOARD_PANTHER
...
+	default UART_QMSI_0_NAME
+
+config UART_QMSI_0_BAUDRATE
+	default 1000000
+
+config UART_QMSI_0_HW_FC
+	def_bool y
+
+endif
+
```
Comment (nashif) [inline boards/x86/panther/Kconfig.defconfig]: remove the whole section above

File: ext/drivers/atmel/winc1500/Kconfig [hunk 19]
```diff
@@ -0,0 +1,9 @@
+# Kconfig - Atmel WINC1500 configuration options
```
Comment (nashif) [inline ext/drivers/atmel/winc1500/Kconfig]: everything is commented out here, so what is the point of adding this file?

**Review by nashif — CHANGES_REQUESTED** [review]

File: ext/drivers/atmel/winc1500/Kconfig (hunk 19 above)
Comment (MaxAgneni) [inline ext/drivers/atmel/winc1500/Kconfig]: point to add this file is drivers/wifi/winc1500/Kconfig

Comment (nashif) [inline ext/drivers/atmel/winc1500/Kconfig]: but the while thing is commented, so it has no effect at all

Comment (MaxAgneni) [inline ext/drivers/atmel/winc1500/Kconfig]: in Makefile there is:
obj-$(CONFIG_WIFI_WINC1500) += ....

File: boards/x86/panther/Kconfig.defconfig (hunk 18 above)
Comment (MaxAgneni) [inline boards/x86/panther/Kconfig.defconfig]: This part of Kconfig is just removed

**Comment by drahnr** [comment]
Is there a general need for the offload ip stack with this driver? The last commit only removes a couple of dependencies, but not all of them.

**Comment by galak** [comment]
Taken over by #4711

**Pull request closed without merging**